### PR TITLE
Feature/public holiday highlighting

### LINKED ISSUE #107: Weather and Public Holiday API
## User story
As a student, I want to see today's weather and upcoming public holidays on 
my dashboard, so that I can plan consultations around disruptions.

## Context
The brief (section 3.1) allows additional functionality using third-party web 
APIs and mentions RapidAPI as a useful resource. This story adds two lightweight 
API integrations to the student dashboard:

1. **Weather** — current weather for Johannesburg shown in the dashboard header 
   area (temperature + condition icon). Use the Open-Meteo API (free, no key 
   required) with Johannesburg coordinates.

2. **Public holidays** — fetch South African public holidays for the current 
   year from the Nager.Date API (free, no key required). Holidays that fall 
   within the 10-day calendar window should be marked on the calendar cells 
   so students know not to expect consultations on those days.

Both APIs are called server-side in the dashboard controller to avoid CORS 
issues and keep API logic out of the browser.

## Acceptance tests
- [ ] Student dashboard shows current temperature and weather condition for 
      Johannesburg
- [ ] If the weather API is unreachable, the dashboard still loads — the 
      weather widget shows "Unavailable" rather than crashing
- [ ] Public holidays within the 10-day window are marked on the calendar 
      with a label (e.g. "Public holiday") and no Schedule buttons
- [ ] If the holidays API is unreachable, the calendar renders normally 
      without holiday markers — no crash
- [ ] API calls are not made on every request — cache the holiday list for 
      the day using a module-level variable

**Labels:** `user-story` `enhancement`  
**Story points:** 5  
**Sprint:** 4  
**Parent:** #3 (View consultation dashboards)

### PR DESCRIPTION
**Pull request by Aditya-Raghunandan**
## feat(dashboard): weather forecast + public holiday integration

**Story:** #107
**Sprint:** 4  
**Branch:** `feature/public-holiday-highlighting`

---

### Summary

Two lightweight server-side API integrations that surface contextual
information at the point where students make booking decisions. No new
npm packages, no schema changes, no API keys required.

---

### What ships in this PR

**South African Public Holidays — Nager.Date API**

- Student 10-day calendar: holiday days render with amber highlight,
  availability slots are hidden, Schedule/Join buttons are replaced with
  a "Public holiday" badge
- Lecturer monthly calendar: holiday cells highlighted amber with a
  Bootstrap tooltip showing the holiday name
- If no holidays fall in the visible window: small on-theme message
  renders below the filter toggle —
  *"🚪 No public holidays in the next 10 days — all doors are fully
  operational."*
- API unreachable → calendar renders exactly as before, no crash

**Wits Weather Forecast — Open-Meteo API**

Three surfaces, each with different behaviour depending on condition:

*Student 10-day calendar (informational only):*

Every day column header shows a weather icon next to the date. Hovering
shows the full condition message and max temperature via Bootstrap
tooltip. No booking behaviour changes here — the icon is purely
contextual.

```
Mon, 12 May ☀️   ← hover: "Clear skies · 24°C"
Tue, 13 May 🌧️  ← hover: "Rain expected — bring an umbrella · 18°C"
```

*Lecturer monthly calendar (informational only):*

Each date cell shows the weather icon on a second line at 0.7rem with a
Bootstrap tooltip. Holiday amber cells and today/has-consultation
classes are unaffected.

```
┌────┐
│ 11 │
│ ☀️  │  ← hover: "Clear skies · 24°C"
└────┘
```

*Booking page (condition-driven behaviour):*

The meta line gets the icon and max temp appended next to the date.
Above the Title input, a contextual banner renders depending on
condition:

| Condition | Banner |
|---|---|
| `sunny`, `cloudy`, `mild` | No banner — meta line only |
| `rainy` | `alert-info` 🌧️ — *"Rain is expected on this date. If you'd prefer an online session, mention it in the description to request a Teams link."* |
| `storm` | Same `alert-info` banner as rainy |
| `cold` (WMO 71–77) **or** any condition where maxTemp ≤ 14°C | `alert-warning` 🧥 — *"It'll be chilly on this date. Bundle up — or suggest an online session in the description."* |

The cold check is a double-trigger: WMO snow/freezing fog codes fire it,
but so does any condition where max temperature is 14°C or below — so a
clear sunny Joburg winter day at 12°C still renders the chilly banner
even though WMO reports clear skies.

**All conditions — API unreachable or date outside 10-day window:**
the service returns `{}` silently. No icons, no banners, no errors. The
booking form works identically to before this feature existed.

**ADR-011**

Documents both API decisions, alternatives rejected, shared
architectural pattern, and explicit rubric alignment with Section 3.1
of the project brief.

---

### Architecture

Both services follow the same pattern:

| Concern | Approach |
|---|---|
| API calls | Server-side only via Node 18 `fetch` |
| Caching | Module-level — holidays: process lifetime, weather: 4-hour TTL |
| Failure mode | Silent `{}` / `[]` fallback, page renders normally |
| New packages | None |
| Schema changes | None |

---

### Commits

```
83de2d1 test(weather): add weather service unit, integration, and E2E tests
151859c documentation(adr): add ADR-011 for third-party API integration  
4022f9b feat(dashboard): add Wits weather forecast via Open-Meteo API
9e22034 test(dashboard): add public holiday service unit and integration tests
cec30fb feat(dashboard): add SA public holiday highlighting via Nager.Date API
```

---

### Test results

```
Test Suites: 31 passed, 31 total
Tests:       269 passed, 269 total
```

New tests added (all existing tests untouched):

| File | What it covers |
|---|---|
| `tests/unit/weather-service.test.js` | WMO mapping (sunny, rainy, storm), chilly override at ≤14°C, network failure, module-level cache — 7 tests |
| `tests/unit/public-holidays-service.test.js` | Array shape, network failure, cache — 3 tests |
| `tests/integration/studentDashboard.test.js` | Dashboard renders when both services throw; `'Public holiday'` appears when holiday falls in window |
| `tests/integration/consultationBooking.test.js` | Booking page renders when weather throws; rain banner renders when condition is `rainy` |
| `tests/e2e/booking-flow.spec.js` | Booking page body never contains literal `'undefined'` or `'null'` |

---

### How to test locally

```bash
npm run setup
node app.js
```

**Student dashboard** — log in as `1234567` / `pass`. Each day column
header in the 10-day calendar shows a weather icon. Hover for the
tooltip. Click Schedule on any slot — the booking form meta line shows
the icon and temp. If the forecast is rainy an `alert-info` banner
appears above the Title field. If max temp is ≤ 14°C an `alert-warning`
chilly banner appears instead.

**Lecturer dashboard** — log in as `A000356` / `pass`. The monthly
calendar shows small weather icons inside date cells. May 1 (Workers
Day) renders amber if it falls in the current month view.

**Graceful degradation** — disable network in DevTools, reload either
dashboard. Calendars render normally, booking form works, no broken
icons or errors anywhere.

---

### Acceptance tests

- [x] Student 10-day calendar shows weather icon and tooltip for each
      day
- [x] Lecturer monthly calendar shows weather icon inside each date cell
- [x] Booking page meta line shows weather icon and max temp
- [x] Rainy or storm forecast → `alert-info` banner above Title input
- [x] Cold condition or maxTemp ≤ 14°C → `alert-warning` chilly banner
      above Title input (fires on WMO snow/fog codes AND on any sunny
      day below 14°C)
- [x] Sunny, cloudy, mild → no banner, meta line only
- [x] Weather API unreachable or date outside window → no icons, no
      banners, booking form identical to pre-feature behaviour
- [x] Public holidays within the 10-day window show amber highlight and
      hide Schedule buttons
- [x] Lecturer calendar highlights holiday cells amber with tooltip
- [x] "No public holidays" on-theme message renders when window is clear
- [x] Public holidays API unreachable → calendar renders normally
- [x] No literal `'undefined'` or `'null'` rendered anywhere on the
      booking page

---

### Files changed

**New:**
- `src/services/weather-service.js`
- `src/services/public-holidays-service.js`
- `documentation/ADR's/ADR-011-third-party-apis.md`
- `tests/unit/weather-service.test.js`
- `tests/unit/public-holidays-service.test.js`

**Modified:**
- `src/controllers/student-dashboard-controller.js`
- `src/controllers/lecturer-dashboard-controller.js`
- `src/controllers/consultation-booking-controller.js`
- `src/views/student-dashboard.html`
- `src/views/lecturer-dashboard.html`
- `src/views/consultation-new.html`
- `tests/unit/dashboard-controller.test.js`
- `tests/integration/studentDashboard.test.js`
- `tests/integration/consultationBooking.test.js`
- `tests/e2e/booking-flow.spec.js`

---

### Reviewer checklist

- [ ] Weather icons visible on student 10-day calendar headers with
      working tooltips
- [ ] Weather icons visible inside lecturer monthly calendar cells with
      working tooltips
- [ ] Public holiday cells amber on both dashboards
- [ ] Rain/storm `alert-info` banner appears on booking page for
      applicable forecasts
- [ ] Chilly `alert-warning` banner appears for cold WMO codes and for
      any day ≤ 14°C regardless of sky condition
- [ ] Sunny/cloudy/mild days show no banner — meta line only
- [ ] All 269 tests pass locally (`npm test`)
- [ ] Graceful degradation verified with network disabled in DevTools
- [ ] ADR-011 content is accurate and follows existing ADR format
- [ ] Commit messages follow Angular convention

closes #107

### CODE REVIEW

File: src/services/weather-service.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+let cache = null;
+let cacheTime = 0;
+const TTL_MS = 4 * 60 * 60 * 1000;
+
+function mapWMO(code, maxTemp) {
+  let base;
+  if (code <= 1)                                              base = { condition: 'sunny',  icon: '☀️',  message: 'Clear skies' };
+  else if (code <= 3)                                         base = { condition: 'cloudy', icon: '⛅',  message: 'Partly cloudy' };
+  else if ((code >= 51 && code <= 67) || (code >= 80 && code <= 82)) base = { condition: 'rainy',  icon: '🌧️', message: 'Rain expected — bring an umbrella' };
```
Comment (ThandiAlgorithm) [inline src/services/weather-service.js:9]: This line exceeds the 100-character maximum set in Section 11 of the style guide. Maybe try to break it across multiple lines.

File: src/services/weather-service.js
Comment (ThandiAlgorithm) [inline src/services/weather-service.js:1]: Clean and well-structured. The mapWMO function covers the WMO code ranges sensibly. One thing worth noting: WMO codes 2-3 are mapped to cloudy but codes 4–50 (fog, haze, etc.) fall through to the else "mild" catch-all ,that's acceptable for this scope but worth a comment so future maintainers know it's intentional, not an oversight. The 4-hour TTL cache is a good call.

File: src/services/public-holidays-service.js
Comment (ThandiAlgorithm) [inline src/services/public-holidays-service.js:1]: Minimal and correct. The module-level cache object keyed by year is the right pattern for data that doesn't change mid-year. The try/catch returning [ ] on any failure is exactly what the graceful-degradation requirement demands.

File: src/controllers/student-dashboard-controller.js
Comment (ThandiAlgorithm) [inline src/controllers/student-dashboard-controller.js:1]: The Promise.all with .catch(() => fallback) on each call is the right pattern, it ensures one failing API doesn't block the other or crash the render. The conversion from a synchronous to async controller is done correctly.

File: src/controllers/lecturer-dashboard-controller.js
Comment (ThandiAlgorithm) [inline src/controllers/lecturer-dashboard-controller.js:1]: Same Promise.all pattern, same verdict. The publicHolidayDayMap construction loop is clear. Minor style note: _mp as a variable name inside the EJS template is a bit cryptic, monthPadded would read better, but it's a view concern not a logic one.

File: src/controllers/consultation-booking-controller.js
Comment (ThandiAlgorithm) [inline src/controllers/consultation-booking-controller.js:1]: getWitsWeather().catch(() => ({})) is correct. Passing weatherForDate: null when the date isn't in the forecast window is the right safe value, the view guards with if (weatherForDate) throughout.

File: src/views/consultation-new.html
Comment (ThandiAlgorithm) [inline src/views/consultation-new.html:1]: This mirrors the PR description's "double-trigger" intent, a WMO snow code fires it and any day at or below 14°C fires it regardless of sky condition. The logic is correct, but it means a rainy day at 13°C would show the cold banner, not the rain banner, because the else if evaluates the temperature check second. Whether that's the intended priority is worth confirming with the brief, rain at 13°C could reasonably warrant the rain banner instead, or both.

File: documentation/ADR's/ADR-011-third-party-apis.md
Comment (ThandiAlgorithm) [inline documentation/ADR's/ADR-011-third-party-apis.md:1]: Well-written. Covers context, decision, rejected alternatives with concrete reasons, and consequences (including the honest tradeoffs). Consistent with what the code actually does.

File: src/views/lecturer-dashboard.html
Comment (ThandiAlgorithm) [inline src/views/lecturer-dashboard.html:1]: The .holiday CSS class is defined cleanly in the <style> block. Tooltip initialisation via document.querySelectorAll('[data-bs-toggle="tooltip"]') at the bottom covers both the holiday and weather tooltips in one sweep. The legend update (adding the amber public holiday swatch) is a nice UX touch.

File: src/views/student-dashboard.html [hunk 2]
```diff
@@ -210,39 +231,51 @@ <h4 class="mb-0">Find a Consultation</h4>
...
+            <%
+              const holidayName2 = holidayMap[dateStr];
+              const isHoliday2 = !!holidayName2;
+            %>
+            <div class="day-col card p-2 <%= isToday2 ? 'today-cell' : '' %>"
+                 <% if (isHoliday2) { %>style="background-color:#fff3cd; border: 1px solid #ffc107 !important;"<% } %>
+                 data-day-col>
               <div class="fw-semibold small mb-1 text-center" style="font-size:0.8rem;">
                 <%= dayLabel2 %>
+                <% const wx2 = weatherByDay && weatherByDay[dateStr]; if (wx2) { %><span data-bs-toggle="tooltip" title="<%= wx2.message %> · <%= wx2.maxTemp %>°C" style="cursor:default;"><%= wx2.icon %></span><% } %>
```
Comment (ThandiAlgorithm) [inline src/views/student-dashboard.html:243]: Leaning on Bootstrap's existing tooltip system rather than building a custom hover interaction keeps the template lean and consistent with the rest of the project. The cursor:default detail is a small but thoughtful UX touch, it signals to the user that the icon is informational, not clickable.

File: tests/e2e/booking-flow.spec.js
Comment (ThandiAlgorithm) [inline tests/e2e/booking-flow.spec.js:1]: This is the correct approach for an E2E test, it earns its way to the booking page through the actual login flow rather than jumping straight to the URL. That means the session, auth middleware, and redirect chain are all exercised as part of the test.

File: tests/e2e/booking-flow.spec.js [hunk 3]
```diff
@@ -44,3 +44,23 @@ test('student can book a consultation via the 10-day calendar', async ({ page })
...
+  const scheduleBtn = page.locator('[data-testid="schedule-btn"]').first();
+  await expect(scheduleBtn).toBeVisible();
+  await scheduleBtn.click();
+
+  await page.waitForURL('**/consultations/new**');
+  await expect(page.locator('text=Book a Consultation')).toBeVisible();
+
+  const bodyText = await page.locator('body').innerText();
+  expect(bodyText).not.toContain('undefined');
+  expect(bodyText).not.toContain('null');
```
Comment (ThandiAlgorithm) [inline tests/e2e/booking-flow.spec.js:65]: The null assertion will produce a false positive if the page contains any word with "null" as a substring, for example a lecturer named "McNullagh" or any descriptive text containing "annulled", "nullify" etc.

File: tests/integration/consultationBooking.test.js
Comment (ThandiAlgorithm) [inline tests/integration/consultationBooking.test.js:1]: The structure of this test file is clean and well-organised. Mocking at the module boundary rather than patching fetch globally is the right call, it keeps each test isolated and means the mock behaviour is explicit and predictable. The beforeEach resetting and re-establishing the default mock is good defensive practice that prevents state leaking between tests.

File: tests/integration/studentDashboard.test.js
Comment (ThandiAlgorithm) [inline tests/integration/studentDashboard.test.js:1]: This is the most thorough of the integration test files. The getFirstCalendarWeekday helper is a thoughtful addition, rather than hardcoding a date that could become stale, it computes the correct date dynamically relative to today, which means the holiday injection test will still work correctly regardless of when it runs. The two separate describe blocks also have a clear logical separation: the first tests the holidays service in isolation, the second tests both services together.

File: tests/unit/dashboard-controller.test.js
Comment (ThandiAlgorithm) [inline tests/unit/dashboard-controller.test.js:1]: The update from synchronous to async/await across all seven tests is handled consistently and correctly. Without await on the controller call these tests would have passed vacuously before the async work completed, so catching and fixing all of them in one go is exactly the right approach. The mock setup at the top of the file is also clean, both new services are mocked at module level with safe default return values so existing tests are not disrupted by the new async behaviour.

File: tests/unit/public-holidays-service.test.js
Comment (ThandiAlgorithm) [inline tests/unit/public-holidays-service.test.js:1]: This is a well-structured unit test file. The acceptance criteria comment block at the top is exactly what the style guide (Section 7) describes as appropriate , it documents non-obvious design decisions rather than restating what the code does. The use of jest.resetModules() in beforeEach to reset the module-level cache between tests is the correct technique and shows a good understanding of how the service's caching works.

File: tests/unit/weather-service.test.js
Comment (ThandiAlgorithm) [inline tests/unit/weather-service.test.js:1]: This has a clear structure, a well-designed helper, and test cases that map directly to the acceptance criteria in the PR description. The mockDailyResponse factory shows real care for readability, every test below it is concise because the boilerplate is handled once at the top. The jest.resetModules and jest.restoreAllMocks pairing shows a solid understanding of how Jest's module system and the service's internal cache interact. The WMO mapping tests are specific and targeted, and the chilly override test at 12°C with code 0 demonstrates that the author understood the double-trigger logic well enough to write a test that would catch a regression in exactly that edge case. Overall this file sets the quality standard for the rest of the test suite.

Comment (Aditya-Raghunandan) [reply tests/unit/weather-service.test.js:1]: lets goooo

File: tests/unit/public-holidays-service.test.js
Comment (Aditya-Raghunandan) [reply tests/unit/public-holidays-service.test.js:1]: yes indeed it is

**Review by ThandiAlgorithm — APPROVED** [review]
